### PR TITLE
Variables: Make renamed or missing variable section expandable

### LINKED ISSUE #41619: Grafana 8.2.3: MongoDB operators mistaken for Unknown Variables
**What happened**:
While working with MongoDB in Grafana I discovered extreme lag (~12 seconds) when attempting to load the Global Variables page. In some instances even prompting a wait or kill unresponsive page. I noticed all of the MongoDB operators being used in panel queries were flagged as `Unknown Variables` which I suspect leads to the lag every time I load the page.

The screen captured list you see below continues on for 100+ `Unknow Variables`. Likely the cause of the lag....

image

**What you expected to happen**:
I would expect that these Mongo operators would be filtered out from these warnings.

**How to reproduce it (as minimally and precisely as possible)**:
I've provided a sample dashboard JSON that showcases the issue on a much smaller scale
```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "gnetId": null,
  "graphTooltip": 0,
  "id": 2,
  "iteration": 1636653998904,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "datasource": null,
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "align": "auto",
            "displayMode": "auto"
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "showHeader": true
      },
      "pluginVersion": "8.2.3",
      "targets": [
        {
          "rawQuery": true,
          "refId": "A",
          "target": "$dbName.pmc_top.aggregate([\n    {\"$group\": {\n        \"_id\": \"$Name\",\n        \"calls\": {\n            \"$first\": \"$Calls\"\n        },\n        \"totalDur\": {\n            \"$sum\": \"$TotalDurationNs\"\n        },\n        \"avg\": {\n            \"$sum\": \"$AverageNs\"\n        },\n        \"percen\": {\n            \"$sum\": \"$Percentage\"\n        }\n    }},\n    {\"$set\": {\n    \"array\": [\n        {\n          \"Name\": \"$_id\",\n          \"Calls\": \"$calls\",\n          \"Total Duration - ns\": \"$totalDur\",\n          \"Avg - ns\": \"$avg\",\n          \"Percent - %\": \"$percen\"\n        }\n      ]\n    }},\n    {\"$unwind\": {\n        \"path\": \"$array\"\n    }},\n    {\"$replaceRoot\": {\n        \"newRoot\": \"$array\"\n    }}\n]);",
          "type": "table"
        }
      ],
      "title": "Kernel Stats",
      "type": "table"
    }
  ],
  "schemaVersion": 31,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": [
      {
        "allValue": null,
        "current": {
          "selected": false,
          "text": "miperf_asw_babelstream_f32_mi200",
          "value": "miperf_asw_babelstream_f32_mi200"
        },
        "datasource": null,
        "definition": "workload_names.names.aggregate([\n  {\"$group\": {\n    \"_id\": \"$name\"\n  }}\n]);",
        "description": null,
        "error": null,
        "hide": 0,
        "includeAll": false,
        "label": "Database",
        "multi": false,
        "name": "dbName",
        "options": [],
        "query": "workload_names.names.aggregate([\n  {\"$group\": {\n    \"_id\": \"$name\"\n  }}\n]);",
        "refresh": 1,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "type": "query"
      }
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Sample Mongo Dashboard",
  "uid": "CFJMXdc7z",
  "version": 1
}
```

**Anything else we need to know?**:
This issue seems to have come up in the past (i.e. InfluxDB) and I've listed similar issues below...

- Issue #31212 
- Issue #31563
- PR #31811

**Environment**:
- Grafana version: 8.2.3
- Data source type & version: MongoDB (MongoQL)
- OS Grafana is installed on: Ubuntu 20.04 LTS
- User OS & Browser: Edge (also occurs on Chrome)

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
Another attempt to solve #41619. This PR focuses just on solving the challenge explained in the issue by making the section for Renamed and missing variables expandable. By default the section is closed and the process of identifying renamed/missing variables is only done once when you expand the section.

This PR also reports interaction when a user expands the section and if the expansion takes longer than 1s it reports that as another interaction. This way we can better measure the usage of this feature.

### Initial view
image

### When no renamed or missing variables exist
image

### When renamed or missing variables exist
image
 
**Which issue(s) this PR fixes**:
Fixes #41619

**Special notes for your reviewer**:
```json
{
  "__inputs": [],
  "__elements": [],
  "__requires": [
    {
      "type": "grafana",
      "id": "grafana",
      "name": "Grafana",
      "version": "8.3.0-pre"
    },
    {
      "type": "panel",
      "id": "timeseries",
      "name": "Time series",
      "version": ""
    }
  ],
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": null,
  "iteration": 1637224878191,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "description": "",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "title": "$custom",
      "type": "timeseries"
    }
  ],
  "revision": null,
  "schemaVersion": 33,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": [
      {
        "current": {
          "selected": true,
          "text": [
            "1"
          ],
          "value": [
            "1"
          ]
        },
        "hide": 0,
        "includeAll": true,
        "multi": true,
        "name": "custom_renamed",
        "options": [
          {
            "selected": false,
            "text": "All",
            "value": "$__all"
          },
          {
            "selected": true,
            "text": "1",
            "value": "1"
          },
          {
            "selected": false,
            "text": "2",
            "value": "2"
          },
          {
            "selected": false,
            "text": "3",
            "value": "3"
          }
        ],
        "query": "1,2,3",
        "skipUrlSync": false,
        "type": "custom"
      }
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Unknowns toggle",
  "uid": "Ge5xXX57z",
  "variableSettings": {
    "showUnknowns": true
  },
  "version": 1,
  "weekStart": ""
}

```

### CODE REVIEW

File: packages/grafana-ui/src/components/Collapse/CollapsableSection.tsx [hunk 1]
```diff
@@ -5,20 +5,26 @@ import { Icon } from '..';
 import { GrafanaTheme2 } from '@grafana/data';
 
 export interface Props {
-  label: string;
+  label: string | ReactNode;
```
Comment (ashharrison90) [inline packages/grafana-ui/src/components/Collapse/CollapsableSection.tsx:8]: minor nit, i think `ReactNode` includes `string`. but maybe it helps devs understand they can pass string as well? 🤔

File: public/app/features/variables/inspect/VariablesUnknownTable.test.tsx [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  await waitFor(() =>
+    expect(screen.getByRole('heading', { name: /renamed or missing variables/i })).toBeInTheDocument()
+  );
+
+  return { reportInteractionSpy, getUnknownsNetworkSpy, rerender };
+}
+
+describe('VariablesUnknownTable', () => {
+  describe('when rendered', () => {
+    it('it should render the section header', async () => {
```
Comment (ashharrison90) [inline public/app/features/variables/inspect/VariablesUnknownTable.test.tsx:33]: mega minor nit, don't think we need the `it` at the start of every test

**Review by ashharrison90 — APPROVED** [review]
looks great! 🙌 

1 import that definitely needs fixing, other couple of comments are probably personal preference 🤷

File: packages/grafana-ui/src/components/Collapse/CollapsableSection.tsx (hunk 1 above)
Comment (hugohaggmark) [reply packages/grafana-ui/src/components/Collapse/CollapsableSection.tsx:8]: LOL, true. I'll fix

File: public/app/features/variables/inspect/VariablesUnknownTable.tsx [hunk 3]
```diff
@@ -1,63 +1,129 @@
...
+import { CollapsableSection, HorizontalGroup, Icon, Spinner, Tooltip, useStyles, VerticalGroup } from '@grafana/ui';
 import { GrafanaTheme } from '@grafana/data';
-import { UsagesToNetwork } from './utils';
+
+import { VariableModel } from '../types';
+import { DashboardModel } from '../../dashboard/state';
 import { VariablesUnknownButton } from './VariablesUnknownButton';
+import { getUnknownsNetwork, UsagesToNetwork } from './utils';
+import { useAsync } from 'react-use';
+import { reportInteraction } from '../../../../../packages/grafana-runtime';
```
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply public/app/features/variables/inspect/VariablesUnknownTable.tsx:11]: Oh noeees, thanks for catching this!

File: public/app/features/variables/inspect/VariablesUnknownTable.test.tsx (hunk 2 above)
Comment (hugohaggmark) [reply public/app/features/variables/inspect/VariablesUnknownTable.test.tsx:33]: Again, I usually write `then it` but I missed it. Thanks for noticing.

File: packages/grafana-ui/src/components/Collapse/CollapsableSection.tsx (hunk 1 above)
Comment (hugohaggmark) [reply packages/grafana-ui/src/components/Collapse/CollapsableSection.tsx]: Fixed!

File: public/app/features/variables/inspect/VariablesUnknownTable.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply public/app/features/variables/inspect/VariablesUnknownTable.tsx]: Fixed!

File: public/app/features/variables/inspect/VariablesUnknownTable.test.tsx (hunk 2 above)
Comment (hugohaggmark) [reply public/app/features/variables/inspect/VariablesUnknownTable.test.tsx]: Fixed!

**Review by torkelo — APPROVED** [review]
Looks great :tada:

**Comment by belm0** [comment]
I understand the performance concern, but it seems unfortunate to require users to explicitly expand the section to find out if there are unknown variables.  I suspect that most people will never look, and miss the benefit of the system citing something odd with their panel queries.

**Pull request merged**